packages/runtime/container-runtime/src/opLifecycle/outbox.ts

```diff
@@ -168,16 +177,19 @@ export class Outbox {
 	}
 
 	/**
-	 * Detect whether batching has been interrupted by an incoming message being processed. In this case,
-	 * we will flush the accumulated messages to account for that and create a new batch with the new
-	 * message as the first message.
+	 * Confirm that batching has not been interrupted by an incoming message being processed, unless
+	 * we're under a reentrant context.
 	 *
-	 * @remarks - To detect batch interruption, we compare both the reference sequence number
+	 * @remarks - The Outbox assumes that the base for submitted ops (referenceSequenceNumber and index-within-batch for Grouped Batches)
+	 * does not change during batch accumulation, apart from reentrant ops. i.e. any code block that would advance that baseline must
+	 * flush beforehand and consider any ops submitted during it must be considered reentrant. e.g. see ContainerRuntime.process
+	 *
+	 * @privateRemarks - To detect batch interruption, we compare both the reference sequence number
 	 * (i.e. last message processed by DeltaManager) and the client sequence number of the
 	 * last message processed by the ContainerRuntime. In the absence of op reentrancy, this
 	 * pair will remain stable during a single JS turn during which the batch is being built up.
 	 */
-	private maybeFlushPartialBatch(): void {
+	private assertSequenceNumberCoherency(): void {
```

Comment:
The name is a bit misleading, as it has side-effects (under some conditions), but the name suggest it does not.